libyul/optimiser/ReasoningBasedSimplifier.cpp

```diff
@@ -0,0 +1,305 @@
+	{
+		Literal trueCondition = m_dialect.trueLiteral();
+		trueCondition.location = locationOf(*_if.condition);
+		_if.condition = make_unique<yul::Expression>(move(trueCondition));
+	}
+
+	m_solver->push();
+	m_solver->addAssertion(condition != constantValue(0));
+	CheckResult result2 = m_solver->check({}).first;
+	m_solver->pop();
+	if (result2 == CheckResult::UNSATISFIABLE)
+	{
+		Literal falseCondition = m_dialect.zeroLiteralForType(m_dialect.boolType);
+		falseCondition.location = locationOf(*_if.condition);
+		_if.condition = make_unique<yul::Expression>(move(falseCondition));
+		_if.body = yul::Block{};
+	}
+
+	m_solver->push();
+	m_solver->addAssertion(condition != constantValue(0));
+
+	ASTModifier::operator()(_if.body);
+
+	m_solver->pop();
+}
+
+ReasoningBasedSimplifier::ReasoningBasedSimplifier(
+	Dialect const& _dialect,
+	set<YulString> const& _ssaVariables
+):
+	m_dialect(_dialect),
+	m_ssaVariables(_ssaVariables),
+	m_solver(make_unique<smtutil::SMTPortfolio>())
+{
+}
+
+smtutil::Expression ReasoningBasedSimplifier::encodeExpression(yul::Expression const& _expression)
+{
+	return std::visit(GenericVisitor{
+		[&](FunctionCall const& _functionCall)
+		{
+			if (auto const* dialect = dynamic_cast<EVMDialect const*>(&m_dialect))
+				if (auto const* builtin = dialect->builtin(_functionCall.functionName.name))
+					if (builtin->instruction)
+						return encodeEVMBuiltin(*builtin->instruction, _functionCall.arguments);
+			return newRestrictedVariable();
+		},
+		[&](Identifier const& _identifier)
+		{
+			if (
+				m_ssaVariables.count(_identifier.name) &&
+				m_variables.count(_identifier.name)
+			)
+				return m_variables.at(_identifier.name);
+			else
+				return newRestrictedVariable();
+		},
+		[&](Literal const& _literal)
+		{
+			return literalValue(_literal);
+		}
+	}, _expression);
+}
+
+smtutil::Expression ReasoningBasedSimplifier::encodeEVMBuiltin(
+	evmasm::Instruction _instruction,
+	vector<yul::Expression> const& _arguments
+)
+{
+	vector<smtutil::Expression> arguments = applyMap(
+		_arguments,
+		[this](yul::Expression const& _expr) { return encodeExpression(_expr); }
+	);
+	switch (_instruction)
+	{
+	case evmasm::Instruction::ADD:
+		return wrap(arguments.at(0) + arguments.at(1));
+	case evmasm::Instruction::MUL:
+		return wrap(arguments.at(0) * arguments.at(1));
+	case evmasm::Instruction::SUB:
+		return wrap(arguments.at(0) - arguments.at(1));
+	case evmasm::Instruction::DIV:
+		return smtutil::Expression::ite(
+			arguments.at(1) == constantValue(0),
+			constantValue(0),
+			arguments.at(0) / arguments.at(1)
+		);
+	case evmasm::Instruction::SDIV:
+		// TODO - rounding is different in SMT than in EVM
+		break;
+	case evmasm::Instruction::MOD:
+		return smtutil::Expression::ite(
+			arguments.at(1) == constantValue(0),
+			constantValue(0),
+			arguments.at(0) % arguments.at(1)
+		);
+	case evmasm::Instruction::SMOD:
+		// TODO - this might behave differently in SMT than in EVM
```

Comment:
```suggestion
		// TODO - this might behave differently in SMT than in EVM
		break;
```
It definitely *doesn't* behave like ``LT``, does it?
Actually: Shouldn't this be a compiler error without ``[[fallthrough]]``? Apparently not for otherwise empty cases... maybe we should look into hardening compiler options for that...